pySDC/helpers/fieldsIO.py

```diff
@@ -0,0 +1,731 @@
+            assert (
+                sID not in cls.STRUCTS
+            ), f"struct ID already taken by {cls.STRUCTS[sID]}, cannot use it for {registered}"
+            cls.STRUCTS[sID] = registered
+            registered.sID = sID
+            return registered
+
+        return wrapper
+
+    def initialize(self):
+        """Initialize the file handler : create the file with header, removing any existing file with the same name"""
+        assert self.header is not None, "header must be set before initializing FieldsIO"
+        assert not self.initialized, "FieldsIO already initialized"
+
+        with open(self.fileName, "w+b") as f:
+            self.hBase.tofile(f)
+            for array in self.hInfos:
+                array.tofile(f)
+        self.initialized = True
+
+    def setHeader(self, **params):
+        """(Abstract) Set the header before creating a new file to store the fields"""
+        raise NotImplementedError()
+
+    @property
+    def hInfos(self) -> list[np.ndarray]:
+        """(Abstract) Array representing the grid structure to be written in the binary file."""
+        raise NotImplementedError()
+
+    def readHeader(self, f):
+        """
+        (Abstract) Read the header from the file storing the fields.
+
+        Parameters
+        ----------
+        f : `_io.TextIOWrapper`
+            File to read the header from.
+        """
+        raise NotImplementedError()
+
+    @property
+    def hSize(self):
+        """Size of the full header (in bytes)"""
+        return self.hBase.nbytes + sum(hInfo.nbytes for hInfo in self.hInfos)
+
+    @property
+    def itemSize(self):
+        """Size of one field value (in bytes)"""
+        return self.dtype().itemsize
+
+    @property
+    def fSize(self):
+        """Full size of a field (in bytes)"""
+        return self.nItems * self.itemSize
+
+    @property
+    def fileSize(self):
+        """Current size of the file (in bytes)"""
+        return os.path.getsize(self.fileName)
+
+    def addField(self, time, field):
+        """
+        Append one field solution at the end of the file with one given time.
+
+        Parameters
+        ----------
+        time : float-like
+            The associated time of the field solution.
+        field : np.ndarray
+            The field values.
+        """
+        assert self.initialized, "cannot add field to a non initialized FieldsIO"
+        field = np.asarray(field)
+        assert field.dtype == self.dtype, f"expected {self.dtype} dtype, got {field.dtype}"
+        assert field.size == self.nItems, f"expected {self.nItems} values, got {field.size}"
+        with open(self.fileName, "ab") as f:
+            np.array(time, dtype=T_DTYPE).tofile(f)
+            field.tofile(f)
+
+    @property
+    def nFields(self):
+        """Number of fields currently stored in the binary file"""
+        return int((self.fileSize - self.hSize) // (self.tSize + self.fSize))
+
+    def check(self, idx):
```

Comment:
This function is doing a bit more than its name suggests. How about `convertToPositiveIndex`?